server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -701,8 +714,8 @@ public void applyClusterState(ClusterChangedEvent event) {
                     // 3. Snapshots in any other state that have all their shard tasks completed
                     snapshotsInProgress.entries().stream().filter(
                         entry -> entry.state().completed()
-                            || entry.state() == State.INIT && initializingSnapshots.contains(entry.snapshot()) == false
-                            || entry.state() != State.INIT && completed(entry.shards().values())
+                            || initializingSnapshots.contains(entry.snapshot()) == false
```

Comment:
Before we were finalising all snapshots that had all their shards completed here.
This caught the case where a snapshot had just went from `INIT` to `ABORTED` and was finalised in the repository when at the same time we were initialising it in https://github.com/elastic/elasticsearch/pull/38518/files#diff-a0853be4492c052f24917b5c1464003dR413.

With this change we will never finalise a snapshot here that is still being initialised.